// wlanpub/lib/devinfo.js

var dbhd = require("./dboper");
var Schema = dbhd.Schema;

var schema_devInfo = new Schema({   //模板
                                    // index
    userName    : String,
    bUserTest    :String,
    scenarioId  : String,
    groupName   : String,
    devSN       : String,

    // dev info
    devName     : String,
    devType     : String,
    devBrief    : String,
    devBomCode  : String,

    // scenario info
    scenarioName: String,
    shopName    : String,
    redirectUrl : String,
    basicUrl    : String,
    model       : String
});


function DevInfo() {
    this.devInfo = null;
}

/*
*需要创建和场景微服务一样的表 devinfo
*/

DevInfo.prototype.createDevModel = function (mongoConnParas) {
    dbhd.connectMongoose(mongoConnParas);
    schema_devInfo.index({devSN:1});
    devModel.devInfo = dbhd.mongo.model('devinfo', schema_devInfo);
    dbhd.mongo.on('connected', function() {
        console.log('************************connected*********');
    });
    return   dbhd.mongo;
};

/*
 *根据devSN 从devinfo表中获取所需要的用户及场景信息
 */
DevInfo.prototype.getDevInfo = function(devSN,callback){

    var query = {devSN: devSN};
    var oConditionShow = {"_id":0, "__v":0};
    var response = {userName:"",scenarioId:"",scenarioName:"",shopName:"",bUserTest:""};

    devModel.devInfo.findOne(query,oConditionShow, function(err,doc){
        if (!err){
            if(doc)
            {
                response.shopName   = doc.shopName;
                response.userName   = doc.userName;
                response.bUserTest  = doc.bUserTest;
                response.scenarioId = doc.scenarioId;
                response.scenarioName = doc.scenarioName;
                callback(response,"success");
            }
            else
            {
                response.msg = "dev is not found";
                console.warn("[DEVINFO]The dev[%s] is null when get userinfos",devSN);
                callback(response,"fail");
            }

        }
        else{
            response.msg = "mongodb has a problem";
            console.error("[DEVINFO]find dev[%s] failed: %s",devSN, JSON.stringify(err));
            callback(response,"fail");
        }
    });
}


var devModel = module.exports = new DevInfo;
